User edit modal

lyohe opened the pull request:
本家の方に出してしまっていたので作り直し

## 概要
管理画面からユーザーの名前とメールアドレスを編集できるように下記を追加した。
- route
- controller
- client（入力フォームとモーダル）

## controller の処理
名前（name）、変更先のメールアドレス（emailToBeChanged）、変更前のメールアドレス（currentEmail）を受け取り、 currentEmail でユーザーを検索して該当者の名前とメールアドレスを書き換える。

## 調査中
下記の条件で 500 エラーが出ることがあるので原因を調査する。
- 管理画面を開いた状態で controller のプログラムを書き換え
- サーバーが再度リクエストを受け付けるまで待って
- 画面を更新せずに「編集」ボタンで今回追加したモーダルを出して
- 名前とメールアドレスを入力し
- 「実行」ボタンを押すと
下記のようなエラーが出る。
```
Connect
500 Error: ReactDOMServer does not yet support Suspense.
   at ReactDOMServerRenderer.read (/crowi/node_modules/react-dom/cjs/react-dom-server.node.development.js:3387:25)
   at Object.renderToStaticMarkup (/crowi/node_modules/react-dom/cjs/react-dom-server.node.development.js:4004:27)
   at View.renderFile [as engine] (/crowi/node_modules/express-react-views/index.js:69:32)
   at View.render (/crowi/node_modules/express/lib/view.js:135:8)
   at tryRender (/crowi/node_modules/express/lib/application.js:640:10)
   at Function.render (/crowi/node_modules/express/lib/application.js:592:3)
   at ServerResponse.render (/crowi/node_modules/express/lib/response.js:1012:7)
   at actions.index (/crowi/lib/controllers/admin.ts:83:16)
   at Layer.handle [as handle_request] (/crowi/node_modules/express/lib/router/layer.js:95:5)
   at next (/crowi/node_modules/express/lib/router/route.js:137:13)
   at Route.dispatch (/crowi/node_modules/express/lib/router/route.js:112:3)
   at Layer.handle [as handle_request] (/crowi/node_modules/express/lib/router/layer.js:95:5)
   at /crowi/node_modules/express/lib/router/index.js:281:22
   at Function.process_params (/crowi/node_modules/express/lib/router/index.js:335:12)
   at next (/crowi/node_modules/express/lib/router/index.js:275:10)
   at /crowi/lib/middlewares/adminRequired.ts:6:14
   at Layer.handle [as handle_request] (/crowi/node_modules/express/lib/router/layer.js:95:5)
   at trim_prefix (/crowi/node_modules/express/lib/router/index.js:317:13)
   at /crowi/node_modules/express/lib/router/index.js:284:7
   at param (/crowi/node_modules/express/lib/router/index.js:354:14)
   at param (/crowi/node_modules/express/lib/router/index.js:365:14)
   at Function.process_params (/crowi/node_modules/express/lib/router/index.js:410:3)
   at next (/crowi/node_modules/express/lib/router/index.js:275:10)
   at /crowi/lib/middlewares/loginRequired.ts:33:16
   at Layer.handle [as handle_request] (/crowi/node_modules/express/lib/router/layer.js:95:5)
   at trim_prefix (/crowi/node_modules/express/lib/router/index.js:317:13)
   at /crowi/node_modules/express/lib/router/index.js:284:7
   at param (/crowi/node_modules/express/lib/router/index.js:354:14)
```
たぶん依存先のライブラリか何かで SSR で対応してない (ReactDOMServer) のが原因っぽい。

lyohe committed:
add modal outline to admin page

to enable admin to edit user information

lyohe committed:
add alert for testing

一旦セーブ

一旦セーブ

一旦セーブ

一旦セーブ

enable name and email to edit on admin page

delete comments

ゴミが混じってたので消した

lyohe commented:
package.json を変更していないときの package-lock.json は基本的にゴミ。
node や npm の version が違うとこういうことが起きる。

sotarok commented on lib/controllers/admin.ts:
```diff
@@ -523,5 +523,19 @@ export default (crowi: Crowi) => {
+    const { name, emailToBeChanged, currentEmail } = req.form.userEditForm
+
+    try {
+      const user = await User.findUserByEmail(currentEmail)
```
💡 Should find user by userId (To do so, pass target user id via hidden form element)

sotarok replied:
❌ Check if existing user already has an email to be updated?

sotarok commented on client/components/Admin/User/UserEditModal.tsx:
```diff
@@ -0,0 +1,71 @@
+            <Label>
+              現在のなまえ: <code>{user.name}</code>
+            </Label>
+            <Input type="text" name="name" placeholder="Please enter your new user name" onChange={handleUserNameChange} />
```
💡 Should set current user name as `defaultValue `

sotarok commented on client/components/Admin/User/UserEditModal.tsx:
```diff
@@ -0,0 +1,71 @@
+            <Label>
+              現在のメールアドレス: <code>{user.email}</code>
+            </Label>
+            <Input type="email" name="email" placeholder="Please enter your new email address" onChange={handleEmailChange} />
```
💬 Same ahove.

sotarok commented on lib/controllers/admin.ts:
```diff
@@ -523,5 +523,19 @@ export default (crowi: Crowi) => {
+      const user = await User.findUserByEmail(currentEmail)
+      if (!user) throw new Error('User not found')
+      await user.updateName(name)
+      await user.updateEmail(emailToBeChanged)
```
💡 Should update `name` and `email` at the same time.

sotarok commented on client/components/Admin/User/UserTable.tsx:
```diff
@@ -21,10 +21,11 @@ function withPreventDefault(callback) {
   return {
-    handleClickEdit() {
+    handleClickEdit(user) {
       // TODO: Implement
```
💬 OK to delete this comment :)

lyohe commented:
実行ボタンを右揃えにする

sotarok commented on client/components/Admin/User/UserEditModal.tsx:
```diff
@@ -0,0 +1,71 @@
+    <Modal isOpen={isOpen} toggle={toggle}>
+      <ModalHeader toggle={toggle}>ユーザー情報を編集しますか?</ModalHeader>
+      <ModalBody>
+        <Form onSubmit={handleSubmit}>
```
Use `onSubmit` you defined above via `useForm` instead of `handleSubmit`

lyohe committed:
change placeholder to Japanese

lyohe committed:
delete TODO comment

lyohe committed:
add current name and email as default value of form

lyohe committed:
checkout package-lock.json to 79d3b58385b3943779ad0398daed53b79dc142d2

lyohe committed:
add email duplicate check

lyohe committed:
use onSubmit on useForm instead of handleSubmit in UserEditModal

lyohe committed:
close modal when update is success or fail

lyohe committed:
add fetchUsers to edit function

to update users without refresh

lyohe committed:
align right execute button

いったんセーブ

なんで動いてるのか分からん

lyohe commented:
概要欄の「調査中」はいつの間にか再現しなくなってしまった。

sotarok commented on client/components/Admin/User/UserPage.tsx:
```diff
@@ -148,6 +149,32 @@ function useModal<T = any>(initialState: T | {} = {}) {
   ] as const
 }
 
+function useEditUsers(crowi, setSuccess, setFailure, closeUserEditModal, fetchUsers) {
```
Crowi object は AdminContext からとれそう

例:
https://github.com/lyohe/crowi/pull/1/files#diff-83546de5ed049097c1411d11c193eeebR180

sotarok replied:
とおもったら既存コード `useFetchUsers` も同じようにやってんのね ...

lyohe replied:
いったん今のままでどうでしょう！

sotarok commented on client/components/Admin/User/UserPage.tsx:
```diff
@@ -148,6 +149,32 @@ function useModal<T = any>(initialState: T | {} = {}) {
+  const [name, setName] = useState('')
+  const [emailToBeChanged, setEmailToBeChanged] = useState('')
+
+  const edit = async ({ name, emailToBeChanged, id }) => {
```
`edit` だとなんか、何を edit する関数かわからないから、editUserNameAndEmail とかなんかこう、わかる関数のほうがいいなぁとはおもいつつ

lyohe replied:
今後なにかを編集する関数が登場するかもしれませんし、確かにそうですね。
editUserNameAndEmail にします。

sotarok commented on client/components/Admin/User/UserPage.tsx:
```diff
@@ -180,11 +220,25 @@ const UserPage: FC<{}> = () => {
         move={move}
+        setName={setName}
+        setEmailToBeChanged={setEmailToBeChanged}
+        openUserEditModal={openUserEditModal}
```
```suggestion
        openUserEditModal={({user}) => {
          setName(user.name)
          setEmailToBeChanged(user.email)
          openUserEditModal(user)
        }}
```

ってかんじでいけたりしないだろうか (動作未検証)
(そうしたら setName や setEmailToBeChanged を UserTable に渡す必要ないかな〜って

lyohe replied:
できそうなのでやってみます

lyohe replied:
やってみました

sotarok commented on lib/controllers/admin.ts:
```diff
@@ -523,5 +523,22 @@ export default (crowi: Crowi) => {
+    try {
+      const user = await User.findById(id)
+      if (!user) throw new Error('User not found')
+      const emailDuplicateCheck = await User.findUserByEmail(emailToBeChanged)
```
* email を編集せずに post したとき、自分自身の email が発見され、duplicate error になる
* found だったら id が自分と一致してたら通す処理が必要

lyohe replied:
こちらですが、名前とメールアドレスと編集対象ユーザーの _id をフォームで送ったときに

- 同じメールアドレスを持つユーザー（emailDuplicateUser）を探す
- ↑が存在し、かつ編集対象ユーザーの _id が emailDuplicateUser と等しくない場合に Error を throw する
- ↑の場合以外では名前とメールアドレスの更新処理に入る

という形にしました。

sotarok commented on lib/controllers/admin.ts:
```diff
@@ -523,5 +523,22 @@ export default (crowi: Crowi) => {
+      const user = await User.findById(id)
+      if (!user) throw new Error('User not found')
+      const emailDuplicateCheck = await User.findUserByEmail(emailToBeChanged)
+      if (emailDuplicateCheck) throw new Error('Email duplicated')
```
```suggestion
      if (emailDuplicateCheck) {
        throw new Error('Email duplicated')
      }
```

他の箇所もそうなんだけど、既存コードもそうなってるとおもうけどこっちのほうがいい (わかりやすい)

lyohe replied:
修正します！

sotarok commented on client/components/Admin/User/UserPage.tsx:
```diff
@@ -148,6 +149,32 @@ function useModal<T = any>(initialState: T | {} = {}) {
+  const edit = async ({ name, emailToBeChanged, id }) => {
+    try {
+      const { message } = await crowi.apiPost(`/admin/user/edit`, { userEditForm: { name, emailToBeChanged, id } })
+      setSuccess(message)
```
ためしてみたけど、なんか更新後、また、失敗時、メッセージ出てこないんだよなぁ

lyohe replied:
あれ、そうですね。

lyohe replied:
memo:

`fetchUsers()` で `clearStatus(null)` が呼ばれているので、 success や failure の中身はからになっている。

lyohe replied:
memo: なんで見つけたか

メッセージが出ないんだがなぜ？から6つの可能性

1. きちんと API call の後に setSuccess/setFailure が呼ばれてないか想定外の値が入っている
2. setSuccess/setFailure の関数自体がそれを使っているコンポーネントに渡されて呼ばれているか
3. API からの Response がおかしい可能性（まぁ間違ってたけどそこは問題じゃなかった）
4. setSuccess を呼んでいるメソッドが途中で exception （つまり setSuccess まで到達してない
5. 自分が書いている処理以外の既存の要因でメッセージが上書きされる可能性

5 は component を render する前に console.log すれば 5 は分かったよね。
怪しそうなところ見つけたら順番にコメントアウトしていく。

lyohe committed:
mv function name edit -> editUserNameAndEmail

lyohe committed:
fix email duplication check when the same email address sent to server

lyohe committed:
show alert message when success/fail to update user info

lyohe committed:
fix email duplicate check

not throw error when the same email as before is specified as that to be
changed

lyohe committed:
rewrite response message when successfully updated name or email

lyohe committed:
pass function to update states through openUserEditModal

lyohe committed:
fix typo in UserTable.tsx

getHanlders -> getHandlers

lyohe committed:
fix typo in Modal/PageCreateModal.tsx

setPotalName -> setPortalName

sotarok commented on lib/routes/api/admin.ts:
```diff
@@ -33,6 +33,7 @@ export default (crowi: Crowi, app: Express, form) => {
   router.get('/admin/users', Admin.api.user.index)
   router.get('/admin/users.search', Admin.api.usersSearch)
   router.post('/admin/user/invite', csrf, form.admin.userInvite, Admin.api.user.invite)
+  router.post('/admin/user/edit', csrf, form.admin.userEdit, Admin.api.user.edit)
```
↓にあわせて route 定義 `/:id/edit` にするといいかも

lyohe replied:
route 定義を変えました！

sotarok commented on lib/controllers/admin.ts:
```diff
@@ -523,5 +525,25 @@ export default (crowi: Crowi) => {
+      }
+      const emailDuplicateUser = await User.findUserByEmail(emailToBeChanged)
+      if (emailDuplicateUser && !user.equals(emailDuplicateUser)) {
+        throw new Error('Email duplicated')
```
このエラーのとき、admin は理由知りたいかも

lyohe replied:
throw new Error したときのメッセージをそのまま管理画面に出すようにしました。

lyohe committed:
send user._id whose name or email to be changed as request parameter

lyohe committed:
change editUserNameAndEmail function parameter

lyohe committed:
when throw error with message at updating name or email, then display admin page as fail message

lyohe committed:
update name and email at same time in same method

lyohe commented on lib/models/user.ts:
```diff
@@ -233,6 +234,12 @@ export default (crowi: Crowi) => {
     return this.save()
   }
 
+  userSchema.methods.updateNameAndEmail = function (name, email) {
+    this.name = name
+    this.email = email
+    return this.save()
+  }
+
```
@sotarok ここで一つの関数で名前とメールアドレスを同時にアップデートするようにしました。
